lib/srv/app/server_test.go

```diff
@@ -857,19 +872,48 @@ func TestAuthorize(t *testing.T) {
 				RoleAppLabels: test.roleAppLabels,
 			})
 
+			if test.requireTrustedDevice {
+				authServer := s.authServer.AuthServer
+
+				// Create a role that requires a trusted device.
+				requiredDevRole, err := types.NewRole("require-trusted-devices-app", types.RoleSpecV6{
+					Options: types.RoleOptions{
+						DeviceTrustMode: constants.DeviceTrustModeRequired,
+					},
+					Allow: types.RoleConditions{
+						AppLabels: types.Labels{"*": []string{"*"}},
+					},
+				})
+				require.NoError(t, err, "NewRole")
+				requiredDevRole, err = authServer.CreateRole(ctx, requiredDevRole)
+				require.NoError(t, err, "CreateRole")
+
+				// Add role to test user.
+				user := s.user
+				user.AddRole(requiredDevRole.GetName())
+				user, err = authServer.Services.UpdateUser(ctx, user)
+				require.NoError(t, err, "UpdateUser")
+
+				// Refresh user certificate.
+				s.clientCertificate = s.generateCertificate(t, user, s.appFoo.GetPublicAddr(), "" /* awsRoleARN */)
+			}
+
 			if test.cloudLabels != nil {
-				require.NoError(t, s.appServer.c.CloudLabels.Sync(context.Background()))
+				require.NoError(t, s.appServer.c.CloudLabels.Sync(ctx))
 			}
 
 			s.checkHTTPResponse(t, s.clientCertificate, func(resp *http.Response) {
-				require.Equal(t, test.expectedStatus, resp.StatusCode)
-				buf, err := io.ReadAll(resp.Body)
-				require.NoError(t, err)
-				message := s.message
-				if test.message != "" {
-					message = test.message
+				bodyBytes, err := io.ReadAll(resp.Body)
+				assert.NoError(t, err, "reading response body")
+				require.Equal(t, test.wantStatus, resp.StatusCode, "response status mismatch, body:\n%s", bodyBytes)
```

Comment:
Best to avoid reading bodyBytes if there was an error reading it? Safe to continue the test if the status code was wrong, though
```suggestion
				require.NoError(t, err, "reading response body")
				assert.Equal(t, test.wantStatus, resp.StatusCode, "response status mismatch, body:\n%s", bodyBytes)
```